signet mining utility

ajtowns opened the pull request:
Adds `contrib/signet/miner` for mining signet blocks.

Adds `bitcoin-util` cli utility, with the idea being it can provide bitcoin related functionality that does not rely on the ability to access a running node. Only subcommand currently is "grind" which takes a hex-encoded header and grinds its nonce until its nBits is satisfied.

Updates `getblocktemplate` to include `signet_challenge` field, and makes `getblocktemplate` require the signet rule when invoked on the signet change. Removes connectivity and IBD checks from `getblocktemplate` when applied to a test chain (regtest, testnet, signet).

ajtowns committed:
rpc: update getblocktemplate with signet rule, include signet_challenge

ajtowns committed:
rpc: allow getblocktemplate for test chains when unconnected or in IBD

ajtowns committed:
Add bitcoin-util command line utility

ajtowns committed:
contrib/signet: Add script for generating a signet chain

ajtowns commented:
Some examples:

 * Generate blocks indefinitely, at 10 minute intervals, paying block reward to the given address, using bitcoin-util to do multi-threaded proof-of-work: `$ ./contrib/signet/generate.py --cli='./bitcoin-cli' generate 10 --block-time=600 --address="tb1..." --grind-cmd='./bitcoin-util grind'`
 * Generate 10 blocks, at 10 minute intervals, paying block reward to the given address, with timestamps starting at the genesis block, using single-threaded python code to do proof-of-work: `$ ./contrib/signet/generate.py --cli='./bitcoin-cli' generate 10 --block-time=600 --address="tb1..." --backdate 0`
 * Generate blocks indefinitely at 30 minute intervals, but only if there hasn't been a block in the meantime, paying the reward for block height X to the X'th address from the given descriptor: `$ ./contrib/signet/generate.py --cli='./bitcoin-cli' generate -1 --block-time=1800 --descriptor="wpkh(...)#..." --secondary`

Using a "--block-time=600" will get blocks exactly on 10 minute boundaries, with very little variation. You can alternatively use `--mining-time=20` to target spending an average of 20 seconds of cpu time per block, which should get a mainnet-like distribution of block times once difficulty has stabilised, assuming only one miner. 

You can also generate blocks more manually by running:

 * `./bitcoin-cli -signet getblocktemplate '{"rules": ["signet","segwit"]}' | ../contrib/signet/generate.py --cli='./bitcoin-cli' genpsbt --address=tb1...` -- to generate a psbt, encoding the signet txs and the raw block
 * `./bitcoin-cli -signet -stdin walletprocesspsbt` to sign the psbt from stdin, or some other method of signing a psbt
 * `../contrib/signet/generate.py solvepsbt --grind-cmd='./bitcoin-util grind' | ./bitcoin-cli -signet submitblock` to convert the signed psbt into a signet commitment and submit the block

Using the `--cli` option to specify `./bitcoin-cli -conf=/path/to/custom-signet.conf` and having `signet=1`, `signetchallenge` and a custom datadir all specified in the conf file is probably sensible.

MarcoFalke commented on contrib/signet/generate.py:
```diff
@@ -0,0 +1,422 @@
+#!/usr/bin/env python3
+# Copyright (c) 2010 ArtForz -- public domain half-a-node
+# Copyright (c) 2012 Jeff Garzik
+# Copyright (c) 2010-2020 The Bitcoin Core developers
```
```suggestion
# Copyright (c) 2020 The Bitcoin Core developers
```

:eyes:

ajtowns replied:
Hmm, think those are leftover from when the code from test_framework/messages.py etc was copied in rather than imported

MarcoFalke commented on src/bitcoin-util.cpp:
```diff
@@ -0,0 +1,222 @@
+        }
+
+        char* command = argv[1];
+        if (strcmp(command, "grind") == 0) {
```
Any reason to not use our ArgsManager here, so that `-help` documentation will be created?

ajtowns replied:
I don't think ArgsManager supports subcommands that have different suboptions?

MarcoFalke replied:
It would support `-grind=block_header_hex`. Is something else needed here that I am missing?

ajtowns replied:
I was thinking it would make sense for `bitcoin-util` would have subcommands like git, each with their own different options (like `--amend` for `git commit` vs `--continue` for `git rebase`). `bitcoin-util grind` could have a `-jobs=8` argument specifying how many jobs to run in parallel, eg, which presumably wouldn't be appropriate elsewhere.

MarcoFalke replied:
The wallet tool is doing exactly that already with argsman. E.g. it has the `info` subcommand, or a dump subcommand with conditional options like -dumpfile (#19137)

sipa replied:
This seems worth addressing still: https://github.com/bitcoin/bitcoin/pull/19937/commits/e6473d274fe01d58c9eed75bc56415fcce6d63db#r492742101

MarcoFalke commented on src/bitcoin-util.cpp:
```diff
@@ -0,0 +1,222 @@
+
+    if (argc < 2 || HelpRequested(gArgs)) {
+        // First part of help message is specific to this utility
+        std::string strUsage = PACKAGE_NAME " bitcoin-tx utility version " + FormatFullVersion() + "\n\n" +
```
```suggestion
        std::string strUsage = PACKAGE_NAME " bitcoin-util utility version " + FormatFullVersion() + "\n\n" +
```

maflcko reviewed (commented):
Is there a risk in exposing the miner on RPC as well? Bitcoind is running anyway somewhere, and some miners might want to run everything in one process for convenience? No strong opinion.

ajtowns commented:
I don't think there's a risk in doing it over RPC; it's that it puts the wallet code in the middle of two mining codes bits (ie, generate a template, sign, grind proof of work). Merging mining and wallet code seemed a bit ugly, but keeping them distinct means you need a script to combine them for you, so it was easier to put everything in the script. Also seemed easier to add an RPC later if that makes sense, than to remove it if it doesn't make sense.

kallewoof commented on configure.ac:
```diff
@@ -560,6 +561,12 @@ AC_ARG_ENABLE([util-wallet],
   [build_bitcoin_wallet=$build_bitcoin_utils])
 
+AC_ARG_ENABLE([util-util],
+  [AS_HELP_STRING([--enable-util-util],
```
Why is this needed, since there's already --with-util above? (Also, why util-util?)

ajtowns replied:
`--with-utils` gives the bitcoin-cli, bitcoin-tx, bitcoin-util, bitcoin-wallet utils; `--with-util-X` just gives bitcoin-X. Not really needed, (and util-util is certainly weird) but don't see any reason to be inconsistent.

kallewoof commented on src/bitcoin-util.cpp:
```diff
@@ -0,0 +1,222 @@
+    while (!found && header.nNonce < finish) {
+        const uint32_t next = (finish - header.nNonce < 5000*step) ? finish : header.nNonce + 5000*step;
+        do {
+            if (!(UintToArith256(header.GetHash()) > target)) {
```
`operator<=` exists, so using that seems more straightforward.

kallewoof commented on src/bitcoin-util.cpp:
```diff
@@ -0,0 +1,222 @@
+    std::string error;
+    if (!gArgs.ParseParameters(argc, argv, error)) {
+        tfm::format(std::cerr, "Error parsing command line arguments: %s\n", error);
+        return EXIT_FAILURE;
```
Wouldn't it be totally fine if these simply did exit(...) instead of the EXIT_/CONTINUE_ stuff?

ajtowns replied:
Sure? It's just doing the same thing bitcoin-tx does.

kallewoof reviewed (commented):
Tested ACK e077aa7055e71f08a7d2cfbb29702832899fee04

MarcoFalke commented on src/bitcoin-util.cpp:
```diff
@@ -0,0 +1,222 @@
+        return EXIT_FAILURE;
+    }
+
+    // Check for -chain, -testnet or -regtest parameter (Params() calls are only valid after this clause)
```
Could cherry-pick the syle-changes from commit faf0a26711eed9264113463e56b988cf9fe549fd?

jsarenik commented:
I still have no idea how to run a custom signet on this PR's commit f4c6ec1deaf8cf4dbbbc2633886112ebdddb31dc (which has conflicts when I try to rebase it to master locally).

The [documentation at bitcoin.it](https://en.bitcoin.it/wiki/Signet#Custom_Signet) is out of date. Please help me so that I can test.

kallewoof commented:
@jsarenik I've updated the running an issuer section of the docs. The rest are fine I think. Lemme know if not.

jsarenik commented:
@kallewoof have a look at https://github.com/bitcoin/bitcoin/pull/15454 and add the `createwallet` lines to the doc please.

jsarenik commented:
@kallewoof or start all `bitcoind` with `-wallet` command line argument.

jsarenik commented:
Or actually rather `-wallet=""`

jsarenik commented:
@kallewoof Maybe adding a new `bitcoin-cli createwallet` line after starting a `bitcoind` would be the best according to https://github.com/bitcoin/bitcoin/pull/20034 (because `-wallet=""` may change behavior in next releases as far as I understood).

ajtowns commented:
Rebased, with the a bunch of the suggestions applied

kallewoof commented on src/rpc/mining.cpp:
```diff
@@ -434,6 +435,9 @@ static RPCHelpMan getmininginfo()
     obj.pushKV("pooledtx",         (uint64_t)mempool.size());
     obj.pushKV("chain",            Params().NetworkIDString());
+    if (Params().GetConsensus().signet_blocks) {
+        obj.pushKV("signet_challenge",  HexStr(Params().GetConsensus().signet_challenge));
```
Sorry, maybe this should use `signetchallenge` to match the argument (here and below)?

kallewoof commented:
@jsarenik Gave you edit privileges on the Bitcoin Wiki.

jsarenik commented:
@kallewoof Thank you! Now I need to figure out how to run a custom signet…

jsarenik commented:
Tested ACK f2ee4a95b3d9eac7b717f2ed4316dcb5df2d2795

Editing the [Wiki](https://en.bitcoin.it/w/index.php?title=Signet#Custom_Signet) now.

decryp2kanon commented:
Concept ACK f2ee4a9

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -818,6 +824,10 @@ static RPCHelpMan getblocktemplate()
     aRules.push_back("csv");
     if (!fPreSegWit) aRules.push_back("!segwit");
+    if (consensusParams.signet_blocks) {
+        aRules.push_back("!signet");
```
It would be useful to add a comment for non mining experts, such as yours truly, to explain what this does (especially the `!` part, which I guess doesn't mean "not" here). I found the discussion around `!segwit` rather confusing: https://github.com/bitcoin/bitcoin/pull/17946

ajtowns replied:
Added a comment

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -667,11 +667,13 @@ static RPCHelpMan getblocktemplate()
-    if (node.connman->GetNodeCount(CConnman::CONNECTIONS_ALL) == 0)
-        throw JSONRPCError(RPC_CLIENT_NOT_CONNECTED, PACKAGE_NAME " is not connected!");
+    if (!Params().IsTestChain()) {
+        if (node.connman->GetNodeCount(CConnman::CONNECTIONS_ALL) == 0)
```
Could add missing brackets since you're touching this area.

ajtowns replied:
Done

Sjors commented on contrib/signet/generate.py:
```diff
@@ -0,0 +1,420 @@
+            logging.error("Target mining time must be between 1 and 600")
+            return 1
+
+    mi = json.loads(args.bcli("getmininginfo"))
```
This is not Python 3.5.6 friendly:
```
JSON object must be str, not 'bytes'
```

ajtowns replied:
`args.bcli` / `bitcoin_cli` always returns a string now

Sjors commented on contrib/signet/generate.py:
```diff
@@ -0,0 +1,420 @@
+        psbt = generate_psbt(tmpl, reward_spk, blocktime=start)
+        psbt_signed = json.loads(args.bcli("-stdin", "walletprocesspsbt", input=psbt.encode('utf8')))
+        if not psbt_signed.get("complete",False):
+            sys.stderr.write("PSBT signing failed")
```
Dumping the PSBT result is useful here for debugging.

ajtowns replied:
Should dump `psbt` now if debug logging is enabled. That will contain the whole block as well, so it will be long.

Sjors reviewed (commented):
Concept ACK. Using the wiki instructions, I was able to launch my own signet and mine blocks on it!

It's nice to see `bitcoin-util` #14671 come to live.

The commits that touch the node (3d1c2ce3c2684d4435fe17118d2fbdd10d04a2e0, dd62762ff2d4eac87ce12dff0e8f7ac3dd535f39, 0857bf9e6cb23792de94cfd164289d3eb08c608f) look good to me, but I didn't review the bitcoin-util and generate.py script.

I get a compiler warning:
```
test/validation_tests.cpp:68:88: warning: braces around scalar initializer [-Wbraced-scalar-init]
    BOOST_CHECK(signet_params->GetConsensus().signet_challenge == std::vector<uint8_t>{{OP_TRUE}});
                                                                                       ^~~~~~~~~
/usr/local/include/boost/test/tools/old/interface.hpp:83:6: note: expanded from macro 'BOOST_CHECK'
    (P), BOOST_TEST_STRINGIZE( P ), CHECK, CHECK_PRED, _ )
     ^
/usr/local/include/boost/test/tools/old/interface.hpp:68:61: note: expanded from macro 'BOOST_TEST_TOOL_IMPL'
        BOOST_JOIN( BOOST_TEST_TOOL_PASS_PRED, frwd_type )( P, ARGS ),          \
                                                            ^
/usr/local/include/boost/test/tools/old/interface.hpp:51:47: note: expanded from macro 'BOOST_TEST_TOOL_PASS_PRED2'
#define BOOST_TEST_TOOL_PASS_PRED2( P, ARGS ) P
```

@jsarenik in the wiki it's better to redact the _entire_ public key, because idiots like myself might end up copy-pasting their public key into `REDACTED` and end up with an invalid `-signetchallenge` :-)

jsarenik commented:
> @jsarenik in the wiki it's better to redact the _entire_ public key, because idiots like myself might end up copy-pasting their public key into `REDACTED` and end up with an invalid `-signetchallenge` :-)

Thank you @Sjors for feed-back. Wiki edited. Have a look.

0xB10C commented:
I successfully launched my own signet and mined a few hundred blocks on it, too. Haven't had a chance to take a closer look at the code.

decryp2kanon commented:
just a question about MAGIC_BYTES

https://github.com/bitcoin/bitcoin/blob/b46f37ba5ec4fbd2e4c82343fc4f353d7f34837a/src/chainparams.cpp#L342

https://github.com/bitcoin/bitcoin/blob/b46f37ba5ec4fbd2e4c82343fc4f353d7f34837a/test/functional/test_framework/p2p.py#L116

both are same?

jsarenik commented:
> just a question about MAGIC_BYTES
> 
> https://github.com/bitcoin/bitcoin/blob/b46f37ba5ec4fbd2e4c82343fc4f353d7f34837a/src/chainparams.cpp#L342
> 
> https://github.com/bitcoin/bitcoin/blob/b46f37ba5ec4fbd2e4c82343fc4f353d7f34837a/test/functional/test_framework/p2p.py#L116
> 
> both are same?

Yes. In the other endianness it is `0x40CF030A` and for Signet IV this is the network marker.

decryp2kanon commented on contrib/signet/generate.py:
```diff
@@ -0,0 +1,420 @@
+from binascii import unhexlify
+from io import BytesIO
+
+sys.path.insert(0, "../test/functional/")
```
```suggestion
sys.path.insert(0, "../../test/functional/")
```

in my case, it was wrong path.
```
cd ./contrib/signet/
./generate.py --help
```

jsarenik replied:
Yes, the paths are a bit confusing. It is meant to be run from main source root. See for example https://en.bitcoin.it/wiki/Signet:Custom:Script

ajtowns replied:
Should be fixed

decryp2kanon commented:
I am trying to test. could you give an example of this?
```
usage: generate.py generate [-h] [--block-time BLOCK_TIME]
                            [--target-mining-time TARGET_MINING_TIME]
                            [--backdate BACKDATE] [--secondary]
                            [--signcmd SIGNCMD] [--address ADDRESS]
                            [--descriptor DESCRIPTOR] [--grind-cmd GRIND_CMD]
```

jsarenik commented:
> I am trying to test. could you give an example of this?
> 
> ```
> usage: generate.py generate [-h] [--block-time BLOCK_TIME]
>                             [--target-mining-time TARGET_MINING_TIME]
>                             [--backdate BACKDATE] [--secondary]
>                             [--signcmd SIGNCMD] [--address ADDRESS]
>                             [--descriptor DESCRIPTOR] [--grind-cmd GRIND_CMD]
> ```

See https://github.com/bitcoin/bitcoin/pull/19937#issuecomment-696419619 (above)

ajtowns commented:
Rebased to master to pull in #20060 which fixes the compiler warning and #20157 which activates taproot, and made some changes based on feedback above.

kallewoof commented:
Miner now running this version on my end.

Sjors commented:
The commits that touch the node still good to me, thanks for the changes. Mining also still works. The `bitcoin-util` command help text is missing a list of commands. It's also confusing why `signetseednode` is listed as a potential argument.

MarcoFalke commented:
> command help text is missing a list of commands

This can probably be fixed by switching to ArgsMan: https://github.com/bitcoin/bitcoin/pull/19937#discussion_r510721007

ajtowns committed:
contrib/signet: Document miner script in README.md

ajtowns commented:
Bunch of changes:

 * adds contrib/signet/README.md
 * renames contrib/signet/generate.py to contrib/signet/miner
 * changes many of the parameters (`--block-time` and `--mining-time` replaced by `--nbits` and `--poisson`; `--secondary` replaced by `--multiminer`, `--backup-delay` and `--standby-delay`)
 * makes block times predictable
 * adds a `calibrate` subcommand to see how long your cpu takes to mine a block to pick out a value for `--nbits`
 * improves the output so it's easier to tell what's going on

laanwj commented:
I'm a bit in doubt about adding yet another binary just for this. In principle we already have `bitcoin-tx` with for idea being to provide bitcoin related functionality that does not rely on the ability to access a running node. It's named a bit inconvenient to add this functionality I guess but …

ajtowns commented:
@laanwj bitcoin-util could be used for node-free versions of a bunch of rpc calls (see #14671) and/or to avoid introducing a separate binary for managing asmap stuff (see #18573). I think it would make more sense to aim to eventually deprecate `bitcoin-tx` in favour of psbt operations, and instead keep around `bitcoin-cli` for interacting with a node over rpc, `bitcoin-wallet` for doing wallet operations that don't require a node, and `bitcoin-util` for anything that doesn't require a node or a wallet (or a gui).

muxator commented:
Hi,

a very minor nit: 34b8e1742cb756016bae94a2775a0f7d617050b1 introduces  `contrib/signet/generate.py`. The subsequent commit a2ada2db9c16d50c920ddb56e1b82d30b68987fe updates the documentation, while also renaming `contrib/signet/generate.py` → `contrib/signet/miner`.

Shouldn't the first commit be modified to directly introduce `miner` instead of `generate.py`?

Depending on the project policies, it may also make sense to use a single commit to introduce the new script and update the relevant documentation.

ajtowns commented:
@muxator good catch; fixed to introduce the script as `contrib/signet/miner` directly. Don't see any reason to merge the docs into the same commit -- they're already in the same PR so won't be merged separately.

ajtowns commented on src/Makefile.am:
```diff
@@ -661,6 +667,27 @@ bitcoin_wallet_SOURCES += bitcoin-wallet-res.rc
+  $(LIBUNIVALUE) \
+  $(LIBBITCOIN_COMMON) \
+  $(LIBBITCOIN_UTIL) \
+  $(LIBBITCOIN_CONSENSUS) \
```
"grind" needs libconsensus for both `primitives/block.h` for the block header definition and `arith_uint256.h` for converting nbits to a target and comparing the target to a sha256. Since libconsensus isn't linked into bitcoin-cli, that probably rules out `bitcoin-cli --grind=HEADER` as an alternative.

Making `bitcoin-tx -grind HEADER` work seems plausible; should look something like https://github.com/ajtowns/bitcoin/commits/202009-signet-generate-util2tx 

Long term I think `bitcoin-util` is a good idea, but it might be better to introduce it in its own PR, in which case putting the functionality in `bitcoin-tx` in the meantime seems plausible? Thoughts?

kallewoof replied:
`bitcoin-tx --grind` feels weird to me. Grinding is unrelated to transactions. Unless we decide `bitcoin-util` is not desired, I think introducing it here makes more sense, personally.

muxator commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+
+By default, the signet miner mines blocks at fixed intervals with minimal variation. If you want blocks to appear more randomly, as they do in mainnet, specify the --poisson option.
+
+Using the --multiminer parameter allows mining to be distributed amongst multiple miners. For example, if you have 3 miners and want to share blocks between them, specify --multiminer=1/3 on one, --multiminer=2/3 on another, and --multiminer=3/3 on the last one. If you want one to do 10% of blocks and two others to do 45% each, --multiminer=1-10/100 on the first, and --multiminer=11-55 and --mult ...
```
"occassional runs" -> "occasional runs"

ajtowns replied:
Done

muxator commented on contrib/signet/miner:
```diff
@@ -0,0 +1,639 @@
+from binascii import unhexlify
+from io import BytesIO
+
+PATH_BASE_CONTRIB_SIGNET = os.path.abspath(os.path.dirname(__file__))
```
Hi,

I usually run my binaries putting a symlink to them in a directory somewhere in my PATH (let's say `/home/.local/bin`, but it can be anything, really).

When doing this with `miner`, however, it won't work:

```
$ ln -s <base>/contrib/signet/miner /tmp/miner
$ /tmp/miner --help
Traceback (most recent call last):
  File "/tmp/miner", line 25, in <module>
    from test_framework.blocktools import WITNESS_COMMITMENT_HEADER, script_BIP34_coinbase_height # noqa: E402
ModuleNotFoundError: No module named 'test_framework'
```

The reason is that, when computing the absolute path to the directory containing `miner`, links are not followed. This would work if it included an invocation to [`realpath()`](https://docs.python.org/3/library/os.path.html#os.path.realpath), for example:

```diff
diff --git a/contrib/signet/miner b/contrib/signet/miner
--- a/contrib/signet/miner
+++ b/contrib/signet/miner
@@ -18,7 +18,7 @@ import subprocess
 from binascii import unhexlify
 from io import BytesIO
 
-PATH_BASE_CONTRIB_SIGNET = os.path.abspath(os.path.dirname(__file__))
+PATH_BASE_CONTRIB_SIGNET = os.path.abspath(os.path.dirname(os.path.realpath(__file__)))
 PATH_BASE_TEST_FUNCTIONAL = os.path.join(PATH_BASE_CONTRIB_SIGNET, "..", "..", "test", "functional")
 sys.path.insert(0, PATH_BASE_TEST_FUNCTIONAL)
```

On my copy, I also called `abspath()` when computing the final value for `PATH_BASE_TEST_FUNCTIONAL`, but that's just a question of taste, because I prefer `<base>/test/functional` instead of  `<base>/contrib/signet/../../test/functional` in my include path (it's more explicit to debug if something goes wrong).


The final version would be:
```diff
diff --git a/contrib/signet/miner b/contrib/signet/miner
--- a/contrib/signet/miner
+++ b/contrib/signet/miner
@@ -18,8 +18,8 @@ import subprocess
 from binascii import unhexlify
 from io import BytesIO
 
-PATH_BASE_CONTRIB_SIGNET = os.path.abspath(os.path.dirname(__file__))
-PATH_BASE_TEST_FUNCTIONAL = os.path.join(PATH_BASE_CONTRIB_SIGNET, "..", "..", "test", "functional")
+PATH_BASE_CONTRIB_SIGNET = os.path.abspath(os.path.dirname(os.path.realpath(__file__)))
+PATH_BASE_TEST_FUNCTIONAL = os.path.abspath(os.path.join(PATH_BASE_CONTRIB_SIGNET, "..", "..", "test", "functional"))
 sys.path.insert(0, PATH_BASE_TEST_FUNCTIONAL)
 
 from test_framework.blocktools import WITNESS_COMMITMENT_HEADER, script_BIP34_coinbase_height # noqa: E402
```

From my experience with other Python projects, there should not be any undesired side effects, but I am not the expert here.
Do you think supporting this use case makes sense?

Thanks!

ajtowns replied:
Done

ajtowns commented:
Added @muxator's suggestions and updated bitcoin-util-res.rc to match the new version number scheme (and rebased to master to catch up with the new version scheme)

kallewoof reviewed (commented):
Tested ACK 6258e7eb3c9c4cb519cb519d640b73a059b1cc11

muxator commented on contrib/signet/miner:
```diff
@@ -0,0 +1,639 @@
+        header.nTime = i
+        header.nNonce = 0
+        headhex = header.serialize().hex()
+        cmd = args.grind_cmd.split(" ") + [headhex]
```
If `miner calibrate` is called without any additional arguments, there is a stack trace:

```
$ ./miner calibrate 
Traceback (most recent call last):
  File "./miner", line 637, in <module>
    main()
  File "./miner", line 631, in main
    return args.fn(args)
  File "./miner", line 541, in do_calibrate
    cmd = args.grind_cmd.split(" ") + [headhex]
AttributeError: 'NoneType' object has no attribute 'split'
```

One possible way to fix it could be making `--grind-cmd` required for the calibrate command. For `--solvepsbt` and `--generate`, apparently, this is not needed, since `finish_block()` checks for `None` already.

```diff
diff --git a/contrib/signet/miner b/contrib/signet/miner
--- a/contrib/signet/miner
+++ b/contrib/signet/miner
@@ -604,8 +604,8 @@ def main():
         sp.add_argument("--address", default=None, type=str, help="Address for block reward payment")
         sp.add_argument("--descriptor", default=None, type=str, help="Descriptor for block reward payment")
 
-    for sp in [solvepsbt, generate, calibrate]:
-        sp.add_argument("--grind-cmd", default=None, type=str, help="Command to grind a block header for proof-of-work")
+    for (sp, is_required) in [(solvepsbt, False), (generate, False), (calibrate, True)]:
+        sp.add_argument("--grind-cmd", default=None, type=str, required=is_required, help="Command to grind a block header for proof-of-work")
 
     args = parser.parse_args(sys.argv[1:])
```

Behaviour after the proposed change:
```
$ ./miner calibrate 
usage: miner calibrate [-h] [--nbits NBITS] [--seconds SECONDS] --grind-cmd GRIND_CMD
miner calibrate: error: the following arguments are required: --grind-cmd
```

muxator commented on contrib/signet/miner:
```diff
@@ -0,0 +1,639 @@
+        sys.stderr.write("Can only specify one of --nbits or --seconds\n")
+        return 1
+    if args.nbits is not None and len(args.nbits) != 8:
+        sys.stderr.write("Must specify 8 hex digits for --nbits")
```
A trailing `'\n'` at the end of this line would make is consistent with the previous `sys.stderr.write()`:

```diff
diff --git a/contrib/signet/miner b/contrib/signet/miner
--- a/contrib/signet/miner
+++ b/contrib/signet/miner
@@ -520,7 +520,7 @@ def do_calibrate(args):
         sys.stderr.write("Can only specify one of --nbits or --seconds\n")
         return 1
     if args.nbits is not None and len(args.nbits) != 8:
-        sys.stderr.write("Must specify 8 hex digits for --nbits")
+        sys.stderr.write("Must specify 8 hex digits for --nbits\n")
         return 1
 
     TRIALS = 600 # gets variance down pretty low
```

JeremyRubin commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+
+  cd src/
+  CLI="./bitcoin-cli -conf=mysignet.conf"
+  MINER="..contrib/signet/miner"
```
typo here, should be ../contrib!

samsungbiologics replied:
{
  "address": "1LfV1tSt3KNyHpFJnAzrqsLFdeD2EvU1MK",
  "total_received": 2000001157774,
  "total_sent": 1000000957003,
  "balance": 1000000200771,
  "unconfirmed_balance": 0,
  "final_balance": 1000000200771,
  "n_tx": 56,
  "unconfirmed_n_tx": 0,
  "final_n_tx": 56,
  "hasMore": true,
  "txs": [
    {
      "block_hash": "0000000000000000000da4a0dcefe6ccea3a90e951b050164dcaaa3383bc7e60",
      "block_height": 664638,
      "block_index": 1756,
      "hash": "a1a15c5f3e53c81d2a4e1de05c74c7faed74a8c8d7f2036ffe32200ff2f2d4b0",
      "addresses": ["13GEnxfijEf8SzwL1yEFVh5QFbXBYcgRFt"]

JeremyRubin commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+
+To mine the first block in your custom chain, you can run:
+
+  cd src/
```
maybe just make this a one-liner with some ; or && 's?

JeremyRubin commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+
+Instead of calculating a specific nbits value, --min-nbits can be specified instead, in which case the mininmum signet difficulty will be targeted.
+
+By default, the signet miner mines blocks at fixed intervals with minimal variation. If you want blocks to appear more randomly, as they do in mainnet, specify the --poisson option.
```
How can i set this fixed interval? Slightly confused the difference between nbits and the interval itself. It sounds like interval is every 5 minutes try to mine a 25 seconds difficult block? What if I just want "blocks every 30 seconds"?

kallewoof replied:
> $GENERATE --cli="$CLI" "$@" generate --grind-cmd="$GRIND" --address="$ADDR" --nbits=$NBITS --multiminer=$MULTIMINER --ongoing --poisson

is what it looks like on my end. I think you want --ongoing --poisson.

JeremyRubin replied:
I don't think --poisson is the right thing, as --poisson just adds noise?

I guess the parameter is standby delay, which defaults to 0?

Where I think I'm confused is how setting nbits works because we use the next_block_delta to define how long to wait before mining (via sleeping), but then the grind command conceivably takes est_time(nbits) to mine? so it feels like we're double delaying based on nbits?

ajtowns replied:
You can't choose the interval; the difficulty will just get adjusted so the interval levels out at around 10 minutes. What you can target is a difficulty, so that once you reach that difficulty you keep mining a block every 10 minutes so that you stay at that difficulty. If you're not already at that difficulty, you instead choose a shorter/longer interval, so the difficulty is increased/decreased at the next retarget.

If you want to mine a bunch of blocks to get the chain started, then backdating via `--set-block-time` works. You could also use `--nbits=170f1372` (current mainnet nbits) which would make it try to raise the difficulty by mining quickly (though not faster than an average of 2m30).

There's no double delay, because everything operates on the block time field -- ie, last block had timestamp T, next block has timestamp T+INTERVAL. The standby-delay and backup-delay parameters mean you won't start mining the block until T+INTERVAL+DELAY, and you won't finish mining the block until T+INTERVAL+DELAY+GRINDTIME, but those delays don't affect the following block, which will have a timestamp of T+INTERVAL+INTERVAL2 (though again you might not start mining it for a further DELAY, and it won't be finished until GRINDTIME has also passed)

JeremyRubin replied:
Ahhh this makes sense I didn't realize normal difficulty adjustment rules are still used.

This text could go great in the Readme btw :)

Maybe incl a shell script template to backdate 100 blocks on start to seed a usable wallet?

ajtowns replied:
I think just `START=$(date +%s -d"1000 minutes ago")` and `--set-block-time=$START` would work for that -- even with poisson timing you should end up mining 100+ blocks before you catch up with real time. Maybe something like that should just be the default if you're mining the first block of a signet chain?

JeremyRubin reviewed (approved):
Tested ACK; could do more direct review of the code itself.

Few docs nits.

Utility works like a charm to mine blocks!

gruve-p commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+
+Instead of using a single address, a ranged descriptor may be provided instead (via the --descriptor parameter), with the reward for the block at height H being sent to the H'th address generated from the descriptor.
+
+Instead of calculating a specific nbits value, --min-nbits can be specified instead, in which case the mininmum signet difficulty will be targeted.
```
mininmum > minimum

gruve-p commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+You will then need to pick a difficulty target. Since signet chains are primarily protected by a signature rather than proof of work, there is no need to spend as much energy as possible mining, however you may wish to choose to spend more time than the absolute minimum. The calibrate subcommand can be used to pick a target, eg:
+
+  $MINER calibrate --grind-cmd="$GRIND"
+  nbits=1e00f403 for 25s average mining time
```
nbits=1e00f403 > --nbits=1e00f403

ajtowns replied:
That line is an example of output of the command

gruve-p replied:
The formatting looks strange when reading the README.md

ajtowns replied:
Ah, apparently 4 spaces is the standard, not wordpress's 2 spaces. Fixed.

gruve-p commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+
+These steps can instead be done explicitly:
+
+  $CLI -signet getblocktemplate '{"rules": ["signet","segwit"]}' |
+    $MINER --cli="$CLI" genpsbt --address="$ADDR" |
+    $CLI -signet -stdin walletprocesspsbt |
+    jq -r .psbt |
+    $MINER --cli="$CLI" solvepsbt --grind-cmd="$GRIND" |
+    $CLI -signet -stdin submitblock
```
Maybe fix the formatting here also

gruve-p commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+To mine the first block in your custom chain, you can run:
+
+  cd src/
+  CLI="./bitcoin-cli -conf=mysignet.conf"
+  MINER="..contrib/signet/miner"
+  GRIND="./bitcoin-util grind"
+  ADDR=$($CLI -signet getnewaddress)
+  $MINER --cli="$CLI" generate --grind-cmd="$GRIND" --address="$ADDR" --set-block-time=-1
```
Formatting here also

gruve-p commented on contrib/signet/README.md:
```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
+
+    cd src/
+    CLI="./bitcoin-cli -conf=mysignet.conf"
+    MINER="..contrib/signet/miner"
```
MINER="..contrib/signet/miner" > MINER="../contrib/signet/miner"

laanwj commented:
code review ACK f19148a9ddab01af80067982f2639dbee2cdbf52

sipa commented:
Concept ACK. I think it makes sense to have a bitcoin-util tool with subcommands for wallet- and node-free operations (and deprecate bitcoin-tx in favor of PSBT support in bitcoin-util).

gruve-p commented:
ACK https://github.com/bitcoin/bitcoin/pull/19937/commits/f19148a9ddab01af80067982f2639dbee2cdbf52

practicalswift commented:
Concept ACK

`bitcoin-util` makes sense!

laanwj commented on src/rpc/mining.cpp:
```diff
@@ -714,6 +718,13 @@ static RPCHelpMan getblocktemplate()
 
+    const Consensus::Params& consensusParams = Params().GetConsensus();
+
+    // GBT must be called with 'signet' set in the rules for signet chains
```
What happens when `signet` rules are set when calling `getblocktemplate` on a non-signet network, I suppose this is (or should be) an error?

ajtowns replied:
The way I read BIP9 is that by including "signet" in the rules when doing a GBT request, that just indicates you can handle it if signet rules are active -- if they're not active, then there's no reason why you can't handle that too. So adding signet to rules on mainnet should be fine as far as bitcoind is concerned -- it will just return a regular template without any signet stuff, just as if you'd indicated support for .. extension block mining or something.

laanwj replied:
Thanks, clear.

laanwj commented on src/bitcoin-util.cpp:
```diff
@@ -0,0 +1,217 @@
+         * "argc < 0" should be enough to be unreachable, but not
+         * so unreachable that it's optimised away entirely
+         */
+        try { UniValue uv; uv["txid"]; uv.get_str(); uv.write(4); } catch (...) { }
```
I think this is awful. Can someone of the build system people please weigh in here?

fanquake replied:
This works for me:
```diff
diff --git a/src/Makefile.am b/src/Makefile.am
index b974bed66..f37db7bde 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -677,9 +677,9 @@ bitcoin_util_SOURCES += bitcoin-util-res.rc
 endif
 
 bitcoin_util_LDADD = \
-  $(LIBUNIVALUE) \
   $(LIBBITCOIN_COMMON) \
   $(LIBBITCOIN_UTIL) \
+  $(LIBUNIVALUE) \
   $(LIBBITCOIN_CONSENSUS) \
   $(LIBBITCOIN_CRYPTO) \
   $(LIBSECP256K1)
diff --git a/src/bitcoin-util.cpp b/src/bitcoin-util.cpp
index ff6bedec0..f7e670f4e 100644
--- a/src/bitcoin-util.cpp
+++ b/src/bitcoin-util.cpp
@@ -179,16 +179,6 @@ static int CommandLineUtil(int argc, char* argv[])
 
 int main(int argc, char* argv[])
 {
-    if (argc < 0) {
-        /* this nonsense is needed or you get linker errors because
-         * ArgsManager does these ops but for some reason they're not
-         * available.
-         * "argc < 0" should be enough to be unreachable, but not
-         * so unreachable that it's optimised away entirely
-         */
-        try { UniValue uv; uv["txid"]; uv.get_str(); uv.write(4); } catch (...) { }
-    }
-
     SetupEnvironment();
 
     try {
```

ajtowns replied:
Thanks @fanquake !

ajtowns commented:
Rebased to fix the bad library ordering.

Suggest reviewing #20715 which is probably the future for bitcoin-util's argument parsing. I think we could update bitcoin-util to work that way either before or after merging this PR.

bitmastercoin commented:
block this st?

laanwj commented:
code review ACK 595a34dbea01954cb0372b0210d2fd64357a1762

weedcoder commented:
![4txwyb](https://user-images.githubusercontent.com/4971940/104738010-47915980-5745-11eb-934e-7a4f236e18ec.jpeg)

kallewoof commented:
@weedcoder I'd say it is missing a lot of the logic required to do what liquid does, but I do believe this + a permanent elements chain with a 2-way-peg can do a great job as a test-chain for liquid. :)

hebasto commented on src/bitcoin-util.cpp:
```diff
@@ -0,0 +1,207 @@
+#include <stdio.h>
+#include <thread>
+
+#include <boost/algorithm/string.hpp>
```
Not sure if this `include` is required.
